api/capi/src/tensor_filter_single.c

```diff
@@ -374,12 +573,40 @@ static gboolean
 g_tensor_filter_single_invoke (GTensorFilterSingle * self,
     GstTensorMemory * input, GstTensorMemory * output)
 {
-  gboolean status = TRUE;
-  /** TODO: fill this */
+  gboolean status;
+  int i;
+
+  if (G_UNLIKELY (!self->fw) || G_UNLIKELY (!self->fw->invoke_NN))
+    return FALSE;
+  if (G_UNLIKELY (!self->fw->run_without_model) &&
+      G_UNLIKELY (!self->prop.model_file))
+    return FALSE;
 
   /** start if not already started */
-  if (self->started == FALSE)
+  if (self->started == FALSE) {
     status = g_tensor_filter_single_start (self);
+    if (status == FALSE) {
+      return status;
+    }
+    self->started = TRUE;
+  }
+
+  /** Setup output buffer */
+  for (i = 0; i < self->prop.output_meta.num_tensors; i++) {
+    /* allocate memory if allocate_in_invoke is FALSE */
+    if (self->fw->allocate_in_invoke == FALSE) {
+      output[i].data = g_malloc (output[i].size);
+      if (!output[i].data)
+        goto error;
+    }
+  }
+
+  if (self->fw->invoke_NN (&self->prop, &self->privateData, input, output) == 0)
+    return TRUE;
 
-  return status;
+error:
+  if (self->fw->allocate_in_invoke == FALSE)
+    for (i = 0; i < self->prop.output_meta.num_tensors; i++)
+      g_free (output[i].data);
+  return FALSE;
```

Comment:
What if `g_malloc()` is failed in the middle of index? I think i should be the latest index which is successful.

Reply:
I am not sure if I understand correctly. If you are concerned about calling `g_free()` on index `i` which have not been called with `g_malloc`, `g_free` can work with `NULL` as well.
Note: If `g_malloc` is not called, `output[i].data` is NULL.

Reply:
ohh then ok